File: python_src/22.括号生成.py
#
# @lc app=leetcode.cn id=22 lang=python3
#
# [22] 括号生成
#
from typing import *

# @lc code=start
class Solution:
    template: List[str] = ["(", "", ")", ""]
    def generateParenthesis(self, n: int) -> List[str]:
        note: List[List[str]] = [[] for _ in range(n + 1)]

        note[0].append("")
        note[1].append("()")

        for i in range(2, n + 1):
            for j in range(i):
                k = i - j - 1
                for a in note[k]:
                    for b in note[j]:
                        Solution.template[1] = a
                        Solution.template[3] = b
                        note[i].append(''.join(Solution.template))
        return note[n]
                
# @lc code=end

s = Solution()
print([s.generateParenthesis(i) for i in range(1, 4)])

File: rust_src/22.括号生成.rs
/*
 * @lc app=leetcode.cn id=22 lang=rust
 *
 * [22] 括号生成
 */

// @lc code=start
impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
		let mut ans: Vec<Vec<String>> = Vec::with_capacity((n + 1) as usize);

		ans.push(vec![String::from("")]);
		ans.push(vec![String::from("()")]);

		if n < 2
		{
			return ans.remove(n as usize);
		}

		for i in 2..=n
		{
			let mut v = vec![];
			for j in 0..i
			{
				for s1 in &ans[j as usize]
				{
					for s2 in &ans[(i - j - 1) as usize]
					{
						let mut s = String::from("(");

						s.push_str(s1);
						s.push(')');
						s.push_str(s2);

						v.push(s);
					}
				}
			}
			ans.push(v);
		}

		return ans.remove(n as usize);
    }

}
// @lc code=end

